### PR TITLE
Add dependencies/references support

### PR DESCRIPTION
**Pull request by libretto**
Protobuf references support:
Add references to Protobuf schemas,
Support compare protobuf schemas with references,
Support serialization/deserialization of protobuf schemas with dependencies.

### CODE REVIEW

**Comment by libretto** [comment]
There is some "copyright" lint check which is not passed on github but locally it is passed. Question is what this check means?

**Comment by libretto** [comment]
> I run this using `pre-commit run --all-files`.
in my case pylint ignore copyright rules by some reason because locally I see
(py311) serge@dell:~/workspace/instaclustr/novage_office/karapace_mainstream$ pre-commit run --all-files
trim trailing whitespace.................................................Passed
fix end of files.........................................................Passed
debug statements (python)................................................Passed
copyright................................................................Passed

**Comment by libretto** [comment]
Debian 11 pre-commit rule not works probably because of this error:  
(py311) serge@dell:~/workspace/instaclustr/novage_office/karapace_mainstream$ git grep -ELm1 'Copyright \(c\) 20[0-9]{2} Aiven' -- '*.py' ':!*__init__.py'
error: unknown switch `m'

File: karapace/schema_models.py
Comment (RyanSkraba) [inline karapace/schema_models.py:157]: This method body looks equivalent to using `self._schema_cached = parse(...)`.  Instead of copying it here, we could just use that method.

Is it necessary to push the parsed schema into the base here?

File: karapace/schema_reader.py
Comment (RyanSkraba) [inline karapace/schema_reader.py:45]: Redefined from `karapace.typing`. I can't think of a good reason why some of these are redefined in this file!

File: karapace/schema_reader.py
Comment (RyanSkraba) [inline karapace/schema_reader.py:535]:
```suggestion
```
Leftover debug code?

File: karapace/schema_reader.py
Comment (RyanSkraba) [inline karapace/schema_reader.py:554]: This won't be thrown if the `reference.version` is not present in the subject_data -- you mean `subject_data.get(reference.version, None)`?

File: karapace/schema_references.py
Comment (RyanSkraba) [inline karapace/schema_references.py:20]: Are these used anywhere?

File: karapace/schema_models.py
Comment (RyanSkraba) [inline karapace/schema_models.py:194]:
```suggestion
            isinstance(other, TypedSchema)
```

TypedSchema is the base class for both.

File: karapace/schema_reader.py
Comment (RyanSkraba) [inline karapace/schema_reader.py:504]: Is it intentional that the exceptions raised in `_resolve_reference` are propagated instead of just returning like `InvalidSchema` exceptions?

**Review by RyanSkraba — COMMENTED** [review]
Just some quick notes, so you know that we are looking at this!

I'm not asking for any changes yet -- I like to think a review is a collaboration and I still have quite a bit to get through.  Don't hesitate to let me know if I've read or misunderstoon, and I'll keep on going through!

File: karapace/in_memory_database.py
Comment (libretto) [reply karapace/in_memory_database.py:33]: Yes it may be replaced with Dict[(Subject,ResolvedVersion)]

File: karapace/errors.py
Comment (libretto) [reply karapace/errors.py:54]: ok

File: karapace/utils.py
Comment (libretto) [reply karapace/utils.py:69]: ok

File: karapace/in_memory_database.py
Comment (libretto) [reply karapace/in_memory_database.py:271]: ok

File: karapace/schema_models.py
Comment (libretto) [reply karapace/schema_models.py:285]: No, the superclass has no parsed schema object inside.

Comment (libretto) [reply karapace/schema_models.py:285]: ok

File: karapace/schema_models.py
Comment (libretto) [reply karapace/schema_models.py:157]: will fix

File: karapace/schema_reader.py
Comment (libretto) [reply karapace/schema_reader.py:45]: ok

File: karapace/schema_reader.py
Comment (libretto) [reply karapace/schema_reader.py:535]: yes

File: karapace/schema_reader.py
Comment (libretto) [reply karapace/schema_reader.py:553]: yes

File: karapace/schema_reader.py
Comment (libretto) [reply karapace/schema_reader.py:504]: No, we need to catch InvalidReferences there and log it with a return.

File: karapace/schema_models.py
Comment (libretto) [reply karapace/schema_models.py:194]: yes

File: karapace/schema_references.py
Comment (libretto) [reply karapace/schema_references.py:20]: no
will fix

**Comment by libretto** [comment]
fixed all issues above

File: karapace/schema_models.py
Comment (jjaakola-aiven) [inline karapace/schema_models.py:162]: This  is will cause heavy memory usage as `TypedSchema` is used in the in memory database and must be changed.
The usage is here: https://github.com/aiven/karapace/blob/main/karapace/schema_reader.py#L477-L489

Very early implementation also stored the schemas as objects and caused issues on low memory environments. The schema should be in object format only in the `ParsedTypedSchema` and `ValidatedTypedSchema`.

**Review by jjaakola-aiven — CHANGES_REQUESTED** [review]

File: karapace/schema_models.py
Comment (RyanSkraba) [reply karapace/schema_models.py:162]: Oh, this makes sense: even the name `ParsedTypedSchema` only makes sense if it's the one holding the schema (object).  

@libretto Did you choose to cache the schema in the `TypedSchema` just for performance reasons to avoid reparsing?  This chould be revisited as a future improvement.

Comment (libretto) [reply karapace/schema_models.py:162]: @RyanSkraba Caching was suggested to my code by @jjaakola-aiven at https://github.com/aiven/karapace/commit/984d794b73f308b762965eea3ddabbcd2e5116a3 and I follow his way. Let me try to fix it.

Comment (jjaakola-aiven) [reply karapace/schema_models.py:162]: Yes, it is the commit referenced. It is a mistake from me. The issue is when schema-property is called from the object used in storage that the memory use will creep in the background. I think the compatibility check would have been manifesting this problem in some point, especially transitive which iterates all previous schemas. The current implementation creates always a new `ParsedTypedSchema` object for compatibility check and caching would not even be effective: https://github.com/aiven/karapace/blob/main/karapace/schema_registry.py#L361-L363

Comment (libretto) [reply karapace/schema_models.py:162]: @jjaakola-aiven actually You were right there at this point because You also created the property 'schema'  that will utilize caching in case of multiple calls to this property. But if the schema property is called only once caching will be useless. So it was the right solution but if You know about the memory consumption issue probably we must get rid of it or improve it.

Comment (libretto) [reply karapace/schema_models.py]: @RyanSkraba I just moved cashing to ParsedTypedSchema hope there will be less memory issues

**Comment by aiven-anton** [comment]
@libretto I logged a separate issue for the false positive in the copyright pre-commit hook: https://github.com/aiven/karapace/issues/563

**Comment by RyanSkraba** [comment]
Thanks for the merge with master -- I did this locally and it wasn't too difficult, but I appreciate that it's constant effort. I should have another round of review for you in the protobuf area very soon!

**Review by RyanSkraba — CHANGES_REQUESTED** [review]
Hey,

I've finally gotten all of this code pretty thoroughly reviewed, and again, thanks for your patience!  For me, this is a pretty good feature to ramp up on my Karapace and get some more modern python experience, so I appreciate it.

I really tried to limit myself to noting things that might be _breaking_ changes, since the priority must be to avoid any regressions with customers not using this feature.

There's a couple of things introduced in this PR that could be _improved_, but I think it's wise to create issues and fix them incrementally!  I'm willing to take on some of this work and I hope you'll stick around to help review future changes in this area :D 

Just for discussion: some of the things that I would like to see applied on top of this PR:

- Restricting the scope as tight as possible helps reviews because we can tell the scope where a method applies.

- Some field, file and class names aren't very helpful.  While this often applies to Karapace as a whole, we specifically end up with `dependency.schema.schema` constructions which is hard on the eyes.

- and of course, more extensive unit testing to catch regressions earlier.

To be clear, I'm not looking for the magic PR that makes everybody perfectly happy today!  If we can do a round-trip on these comments, that might be enough to allow the above things to be gradually fixed while providing the references feature to those looking forward to it immediately.

@all, What do you think?

File: karapace/protobuf/proto_file_element.py
Comment (aiven-anton) [reply karapace/protobuf/proto_file_element.py:16]: Double underscore makes it inconvenient to unit test, I reserve them for cases where I'm certain things will unequivocally break if a name is used outside its namespace. For the common case, I'd recommend single underscores.

File: karapace/protobuf/known_dependency.py
Comment (libretto) [reply karapace/protobuf/known_dependency.py:20]: No, but this info may be used as a reference in case we need to regenerate KnownDependency class which is generated by google protobuf predefined(known) types.

File: karapace/protobuf/schema.py
Comment (libretto) [reply karapace/protobuf/schema.py:10]: ok

File: karapace/protobuf/schema.py
Comment (libretto) [reply karapace/protobuf/schema.py:123]: ok

File: karapace/protobuf/compare_type_storage.py
Comment (libretto) [reply karapace/protobuf/compare_type_storage.py:101]: ok

File: karapace/protobuf/io.py
Comment (libretto) [reply karapace/protobuf/io.py:59]: Yes must be schema.schema.

File: karapace/protobuf/proto_file_element.py
Comment (libretto) [reply karapace/protobuf/proto_file_element.py]: will use single underscore

**Comment by libretto** [comment]
@RyanSkraba I fixed all issues from Your comments. Most tests I ported from Wire (a java protobuf parser) or inspired by the code of Schema Registry. Of course, they do not cover all issues. You show an issue with an empty package name where one letter name will raise an exception. I will add there a simple test for one-letter names.

**Comment by RyanSkraba** [comment]
Hello!  Thanks again for doing yet another merge with master -- it looks like there were a couple of hiccups with the merge with formatting and whitespace (which is now a lint error).  

The good news is that I think that once the lint is fixed, this LGTM sufficiently to merge it and make any other requested fixes iteratively!

File: karapace/schema_registry.py
Comment (RyanSkraba) [inline karapace/schema_registry.py:282]:
```suggestion
            ret["references"] = [reference.to_dict() for reference in schema.references]
```
This is a consequence of the JSON-tightening in #570 ... is this the right way to do this?

File: karapace/schema_models.py
Comment (RyanSkraba) [inline karapace/schema_models.py:110]:
```suggestion
        return json_decode(self.schema_str, Dict[str, Any])
```
I think this was a mistake from the merge from main.

File: karapace/schema_registry.py
Comment (libretto) [reply karapace/schema_registry.py:282]: I already applied these fixes locally but integration tests failed... The latest main fixes break part of protobuf logic... probably because JsonObject or something... will debug it ASAP.

**Comment by libretto** [comment]
@RyanSkraba I suppose we waiting for @jjaakola-aiven  approval now...

**Comment by libretto** [comment]
@jjaakola-aiven could I see "changes requested" in this thread all changes are done and all code comments are outdated. Could You take a look at Your change request and approve it?

**Review by libretto — COMMENTED** [review]
All requested changes were fixed.

**Comment by RyanSkraba** [comment]
Thanks so much, all is well, I was doing some follow-up in the meantime -- let's merge this before `main` moves again!  If we have any code style preferences, extra clean-up or fixes we can make those changes incrementally!

**Pull request merged**
